mkosi/__init__.py

```diff
@@ -3398,7 +3453,141 @@ def boot_directory(state: MkosiState, kver: str) -> Path:
     return Path("boot") / state.machine_id / kver
 
 
-def run_kernel_install(state: MkosiState, cached: bool) -> None:
+def kernel_modules_initrd_compression(config: MkosiConfig) -> Union[bool, str]:
+    return should_compress_output(config) or "zstd"
+
+
+def find_compression(path: Path) -> Union[str, None]:
+    with open(path, "rb") as f:
+        magic = f.read(6)
+    if magic[:4] == MAGIC_NUMBERS["zstd"]:
+        return "zstd"
+    elif magic == MAGIC_NUMBERS["xz"]:
+        return "xz"
+    return None
+
+
+def list_kernel_modules_in_initrds(state: MkosiState) -> dict[str, Path]:
+    kernel_modules: dict[str, Path] = {}
+    for initrd in state.config.initrds:
+        try:
+            compression = find_compression(initrd)
+            if not compression:
+                raise MkosiException("Unsupported compression")
+            with subprocess.Popen(uncompress_command(compression, initrd), stdout=subprocess.PIPE) as archive:
+                output: str = run(["cpio", "-it"],
+                                  stdin=archive.stdout, stdout=subprocess.PIPE, stderr=subprocess.DEVNULL, check=True, text=True).stdout
+                for s in output.split("\n"):
+                    path = Path(s)
+                    if path.parent == Path("usr/lib/modules"):
+                        kernel_modules[path.name] = initrd
+        except Exception as err:
+            MkosiPrinter.info(f"Couldn't list kernel modules in {initrd}: {err=}")
+    return kernel_modules
+
+
+@dictify
+def read_os_release(root: Path) -> Iterator[tuple[str, str]]:
+    try:
+        f = root.joinpath('etc/os-release').open()
+    except FileNotFoundError:
+        f = root.joinpath('usr/lib/os-release').open()
+
+    for line_number, line in enumerate(f, start=1):
+        if not line.strip() or line.startswith('#'):
+            continue
+        if m := re.match(r'([A-Z][A-Z_0-9]+)=(.*)', line):
+            name, val = m.groups()
+            if val and val[0] in '"\'':
+                val = ast.literal_eval(val)
+            yield name, val
+        else:
+            print(f'Bad line {line_number}: {line}', file=sys.stderr)
+
+
+def update_suffixed(items: dict[str, str], name: str, fallback: str) -> None:
+    value = items.get(name, fallback)
+    if 'mkosi-initrd' in value:
+        return
+    items[name] = value + ' (mkosi-initrd)'
+
+
+def make_initrd_release(root: Path, out: TextIO) -> None:
```

Comment:
Let's inline this function into `write_initrd_release()` since it's the only caller of this function